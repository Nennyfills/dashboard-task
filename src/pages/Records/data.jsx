import currencyFormat, { dateFormat } from 'helpers';

export const labels = [
  { color: '#A259FF', label: 'Project 1' },
  { color: '#6497B1', label: 'Project 2' },
  { color: '#FFC107', label: 'Project 3' },
  { color: '#F24E1E', label: 'Project 4' }
];

export const chartData = {
  datasets: [
    {
      label: '# of Votes',
      data: [15, 15, 10, 60],
      backgroundColor: [
        '#F24E1E',
        '#FFC107',
        '#6497B1',
        '#A259FF',
      ],
    },
  ],
};
export const tableSchema = [
  {
    title: 'Date',
    render: (rowData) => <span>{dateFormat(rowData?.created) || 'N/A'}</span>,
  },
  {
    title: 'Gateway',
    render: (rowData) => <span>{rowData.gatewayId || 'N/A'}</span>,
  },
  {
    title: 'Transaction ID',
    render: (rowData) => <span>{rowData.paymentId || 'N/A'}</span>,
  },
  {
    title: 'Amount',
    render: (rowData) => <span>{currencyFormat(rowData?.amount)}</span>,
  },
];

export const tableSchema2 = [
  {
    title: 'Date',
    render: (rowData) => <span>{dateFormat(rowData?.created) || 'N/A'}</span>,
  },
  {
    title: 'Transaction ID',
    render: (rowData) => <span>{rowData.paymentId || 'N/A'}</span>,
  },
  {
    title: 'Amount',
    render: (rowData) => <span>{currencyFormat(rowData?.amount)}</span>,
  },
];
