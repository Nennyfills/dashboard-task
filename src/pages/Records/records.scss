.records {
 &--empty {
  // @include center;
  // margin: auto;
  // display: block;
 }
 .label {
  &--inner {
   @include display-flex($align: center);
   margin-right: 1em;

   .color {
    width: 15px;
    height: 15px;
    border-radius: 5px;
    margin-right: .2em;
   }
  }
  @include display-flex($justify: revert, $align: flex-start);
  flex-wrap: wrap;
  margin-right: 1em;
 }
 &__content {
  @include display-flex($align: flex-start);
  @include respond(tab-land) {
   display: block;
  }
  &--width {
   &.override {
    width: 100%;
   }
   width: 60%;
   @include respond(tab-land) {
    width: 100%;
   }
  }
  &__block {
   margin-top: 27px;
   background: $primary-background;
   border-radius: 10px;
   padding: 2em;
   &--header {
    margin-bottom: 2em;
   }
  }
  &--footer {
   margin: 27px 4px 17px;
   padding: 1.3em 2em;
   background: $primary-background;
   border-radius: 10px;
   p {
    text-transform: uppercase;
   }
  }
  &--chart {
   width: 37%;
   @include respond(tab-land) {
    width: 100%;
   }
   &--width {
    height: 24em;
    @include respond(tab-land) {
     height: 40em;
    }
    @include display-flex($justify: center);
    &--inner {
     width: 50%;
    }
   }
  }
 }
}
