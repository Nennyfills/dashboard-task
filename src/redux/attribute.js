import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import httpWrapper from 'api/axiosWrapper';

export const getAllData = createAsyncThunk('report/getAll', async ({ url, id, filterPayload }) =>{
  const res = await httpWrapper({ url });
  if (res) {
    const calls = res?.data?.map(async ()=> await httpWrapper({
      url: 'report',
      payload: filterPayload,
      method: 'post',
    }));
    const result = await Promise.all(calls);
    const allProject = result.map((val, index)=> {
      val.name = res?.data[index]?.name;
      val.id = res?.data[index]?.[id];
      val.total = val?.data?.reduce((current, next)=> next.amount + current, 0);
      return val;
    });
    return allProject;
  }
  return [];
});

export const getReport = createAsyncThunk('report/getReport', async ({ filterPayload, title }) =>{
  const eachReport = await httpWrapper({ url: 'report', payload: filterPayload, method: 'post' });
  const data = [{
    name: title,
    data: eachReport?.data,
    total: eachReport?.data?.reduce((current, next)=> next.amount + current, 0)
  }];
  return data;
});

export const getProjects = createAsyncThunk('report/getProjects', async () =>{
  const projects = await httpWrapper({ url: 'projects' });
  return projects?.data || [];
});

export const getGateways = createAsyncThunk('report/getGateways', async () =>{
  const gateways = await httpWrapper({ url: 'gateways' });
  return gateways?.data || [];
});

export const getUser = createAsyncThunk('report/getUser', async () =>{
  const users = await httpWrapper({ url: 'users' });
  return users?.data[0];
});

export const initialState = {
  allProject: [],
  gateways: [],
  projects: [],
  user: {},
  loading: true,
};
export const attributeSlice = createSlice({
  name: 'attributes',
  initialState,
  extraReducers: {
    [getUser.fulfilled]: (state, { payload }) => {
      state.user = payload;
      state.loading = false;
    },
    [getUser.pending]: (state) => {
      state.loading = true;
    },
    [getAllData.fulfilled]: (state, { payload }) => {
      state.allProject = payload;
      state.loading = false;
    },
    [getAllData.pending]: (state) => {
      state.loading = true;
    },
    [getReport.fulfilled]: (state, { payload }) => {
      state.allProject = payload;
      state.loading = false;
    },
    [getReport.pending]: (state) => {
      state.loading = true;
    },
    [getProjects.fulfilled]: (state, { payload }) => {
      state.projects = payload;
      state.loading = false;
    },
    [getProjects.pending]: (state) => {
      state.loading = true;
    },
    [getGateways.fulfilled]: (state, { payload }) => {
      state.gateways = payload;
      state.loading = false;
    },
    [getGateways.pending]: (state) => {
      state.loading = true;
    },
  },
});

export default attributeSlice.reducer;
