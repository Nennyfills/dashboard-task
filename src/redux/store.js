import { configureStore } from '@reduxjs/toolkit';
import Attribute from './attribute';

export default configureStore({ reducer: { Attribute } });
