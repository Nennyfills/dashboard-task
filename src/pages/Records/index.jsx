import { Accordion, Button, DatePicker, EmptyState, Select } from 'components';
import Table from 'components/Table';
import { Chart as ChartJS, ArcElement, Tooltip, Legend } from 'chart.js';
import { Doughnut } from 'react-chartjs-2';
import { useDispatch, useSelector } from 'react-redux';
import {
  getAllData,
  getReport,
  getGateways,
  getProjects,
} from 'redux/attribute';
import { useEffect, useMemo, useState } from 'react';
import currencyFormat from 'helpers';
import { chartData, labels, tableSchema, tableSchema2 } from './data';

ChartJS.register(ArcElement, Tooltip, Legend);

const Records = () => {
  const [projectHeader, setProjectHeader] = useState('All project');
  const [gateWayHeader, setGateWayHeader] = useState('All gateway');
  const [gateWayId, setGateWayId] = useState(null);
  const [filterPayload, setFilterPayload] = useState({});
  const [projectId, setProjectId] = useState(null);
  const [isGenerating, setIsGenerating] = useState(false);
  const [activeProject, setActiveProject] = useState({ value: 'select', label: 'Select project' });
  const [activeGateway, setActiveGateway] = useState({ value: 'select', label: 'Select gateway' });

  const checkHeader =
  projectHeader === 'All project' || gateWayHeader === 'All gateway';
  const checkHeader2 =
  projectHeader === 'All project' && gateWayHeader === 'All gateway';

  const schema = useMemo(() => tableSchema, []);
  const schema2 = useMemo(() => tableSchema2, []);
  const { Attribute } = useSelector(({ Attribute }) => ({ Attribute }));
  const projectOptions = (Attribute?.projects?.map((itm) =>
    ({ label: itm?.name, value: itm?.projectId })) || []);
  const gatewayOptions = (Attribute?.gateways?.map((itm) =>
    ({ label: itm?.name, value: itm?.gatewayId })) || []);

  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getAllData({ url: 'projects', id: 'projectId', filterPayload }));
    dispatch(getProjects());
    dispatch(getGateways());
  }, []);

  useEffect(() => {
    if (Attribute?.allProject?.[0]?.data?.length) {
      setActiveProject({ value: 'project', label: 'All project' });
      setActiveGateway({ value: 'gateway', label: 'All gateway' });
    }
  }, [Attribute?.allProject?.[0]?.data]);

  const handleOnchange = ({ target }) => {
    setFilterPayload({
      ...filterPayload,
      [target.id]: target?.value,
    });

    if (target?.options[target?.selectedIndex]?.text === 'All project') {
      dispatch(getAllData({ url: 'projects', id: projectId, filterPayload }));
    }
    if (target?.options[target?.selectedIndex]?.text === 'All gateway') {
      dispatch(getAllData({ url: 'gateways', id: gateWayId, filterPayload }));
    } else {
      dispatch(
        getReport({
          idType: target?.id,
          filterPayload,
          title: target?.options[target?.selectedIndex]?.text,
        })
      );
      setGateWayId(target?.value);
    }

    if (target?.id === 'projectId') {
      setProjectHeader(target?.options[target?.selectedIndex]?.text);
      setProjectId(target?.value);
    } else {
      setGateWayHeader(target?.options[target?.selectedIndex]?.text);
      setGateWayId(target?.value);
    }
  };

  return (
    <div className="records" data-testid="records">
      <section className="page-hero flex">
        <div className="page-hero__left" data-testid="">
          <h2 className="page-hero__left--title">Reports</h2>
          <p className="page-hero__left--desc">
            Easily generate a report of your transactions
          </p>
        </div>
        <div className="flex">
          <Select
            options={projectOptions}
            id="projectId"
            onChange={handleOnchange}
            placeholder="Select project"
            active={activeProject}
          />
          <Select
            options={gatewayOptions}
            id="gatewayId"
            onChange={handleOnchange}
            placeholder="Select gateway"
            active={activeGateway}
          />
          <DatePicker id="from" onChange={handleOnchange} />
          <DatePicker
            id="to"
            onChange={handleOnchange}
            placeholder="To date"
          />
          <Button
            variant="secondary"
            className="isGenerating"
            onClick={() => setIsGenerating(!isGenerating)}
          >
            Generate report
          </Button>
        </div>
      </section>

      {!Attribute?.loading && !Attribute?.allProject?.[0]?.data?.length ? (
        <section className="records--empty">
          <EmptyState />
        </section>
      ) : null}

      {Attribute?.allProject?.[0]?.data?.length ? (
        <section className="records__content">
          <div
            className={`records__content--width ${isGenerating ? '' : 'override'}`}
            data-testid="records-content"
          >
            <div className="records__content__block">
              <p className="bold records__content__block--header">
                {projectHeader}
                {' '}
                |
                {' '}
                {gateWayHeader}
              </p>
              {checkHeader ? (
                <Accordion titleData={Attribute?.allProject} loading={Attribute?.loading}>
                  {Attribute?.allProject?.map(({ data }, idx) => (
                    <div key={idx}>
                      <Table
                        schema={checkHeader2 ? schema : schema2}
                        data={data || []}
                        loading={Attribute?.loading}
                        col={checkHeader2 ? schema?.length : schema2?.length}
                      />
                    </div>
                  ))}
                </Accordion>
              ) : (
                <Table
                  schema={checkHeader2 ? schema : schema2}
                  data={Attribute?.allProject?.[0]?.data || []}
                  loading={Attribute?.loading}
                  cols={checkHeader2 ? schema?.length : schema2?.length}
                />
              )}
            </div>
            {checkHeader2 && (
            <div className="records__content--footer">
              <p className="bold">
                Total:
                {' '}
                {currencyFormat(Attribute?.allProject?.reduce((current, next)=>
                  next.total + current, 0))}
              </p>
            </div>
            )}
          </div>

          {isGenerating && (
          <div className="records__content--chart" data-testid="records__content--chart">
            <div className="records__content--footer label">
              {labels?.map((itm) => (
                <div key={itm?.color} className="label--inner p small">
                  <div className="color" style={{ backgroundColor: itm?.color }} />
                  {itm?.label}
                </div>
              ))}
            </div>
            <div className="records__content--chart--width">
              <div className="records__content--chart--width--inner">
                <Doughnut data={chartData} />
              </div>
            </div>
            <div className="records__content--footer" data-testid="records__content--footer">
              <p className="bold">GATEWAY TOTAL | 14,065 USD</p>
            </div>
          </div>
          )}
        </section>
      ) : null}
    </div>
  );
};

export default Records;
