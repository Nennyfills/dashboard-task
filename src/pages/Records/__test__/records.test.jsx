import { textTender, withoutRender } from 'test/testUtils';

import { attributeSlice } from 'redux/attribute';
import { configureStore } from '@reduxjs/toolkit';
import { screen, waitFor } from '@testing-library/react';
import { allProjectPayload } from 'test/fakeData';
import Records from '..';

const store = configureStore({
  reducer: attributeSlice.reducer,
  initialState: { allProject: allProjectPayload },
});

const getAll = () => ({
  type: 'report/getAll/fulfilled',
  payload: allProjectPayload,
});
const loadEmpty = () => ({
  type: 'report/getAll/fulfilled',
  payload: [],
});

it('Should take a snapshot', () => {
  const { asFragment } = textTender(<Records />, store);
  expect(asFragment(withoutRender(<Records />, store))).toMatchSnapshot();
});

it('Should render records without crashing', () => {
  textTender(<Records />, store);
  const element = screen.getByTestId('records');
  expect(element).toBeInTheDocument();
});
it('Should show records empty card when data is null', () => {
  store.dispatch(loadEmpty());
  textTender(<Records />, store);
  const element = screen.getByTestId('empty-state');
  expect(element).toBeInTheDocument();
});

it('Should show records', async () => {
  store.dispatch(getAll());

  await waitFor(() => {
    expect('Project 1').toEqual(store.getState()?.allProject?.[0]?.name);
  });
  await waitFor(() =>
    expect(400000).toEqual(store.getState()?.allProject?.[0]?.total));
});
