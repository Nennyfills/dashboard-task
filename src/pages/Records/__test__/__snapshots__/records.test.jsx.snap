// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="records"
    data-testid="records"
  >
    <section
      class="page-hero flex"
    >
      <div
        class="page-hero__left"
        data-testid=""
      >
        <h2
          class="page-hero__left--title"
        >
          Reports
        </h2>
        <p
          class="page-hero__left--desc"
        >
          Easily generate a report of your transactions
        </p>
      </div>
      <div
        class="flex"
      >
        <div
          class="select-container"
          data-testid="select-container"
        >
          <select
            data-testid="select projectId"
            id="projectId"
          >
            <option
              value="select"
            >
              Select project
            </option>
          </select>
        </div>
        <div
          class="select-container"
          data-testid="select-container"
        >
          <select
            data-testid="select gatewayId"
            id="gatewayId"
          >
            <option
              value="select"
            >
              Select gateway
            </option>
          </select>
        </div>
        <div
          class="date-picker"
          data-testid="date-picker"
        >
          <input
            data-testid="input from"
            id="from"
            max="2022-05-23"
            placeholder="From date"
            required=""
            type="date"
          />
        </div>
        <div
          class="date-picker"
          data-testid="date-picker"
        >
          <input
            data-testid="input to"
            id="to"
            max="2022-05-23"
            placeholder="To date"
            required=""
            type="date"
          />
        </div>
        <div
          class="btnContainer"
          data-testid="button"
        >
          <button
            class="isGenerating secondary xs"
            data-test="reuseable-button"
            id="disabled"
            type="submit"
          >
            Generate report
          </button>
        </div>
      </div>
    </section>
    <section
      class="records--empty"
    >
      <div
        class="empty-state"
        data-testid="empty-state"
      >
        <div
          class="empty-state--detail"
        >
          <h2>
            No reports
          </h2>
          <p
            class="bold"
          >
            Currently you have no data for the reports to be generated.
          Once you start generating traffic through the Balance application 
          the reports will be shown.
          </p>
        </div>
        <svg>
          noReportIcon.svg
        </svg>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
